### PR TITLE
Beats issue#597 edits

### PR DESCRIPTION
**Pull request by dedemorton**
As a first pass at resolving issue #597 I've edited the content and made quite a few changes because I found the existing descriptions (especially wrt how to the negate and match settings) hard to follow. It was really pretty difficult to come up with descriptions that (hopefully) make everything clear.

As a next step (after someone has reviewed and merged the edits), I'll make structural changes to address the problem of the configuration reference info being *way too long*. In fact, some of this content might go into a separate topic that has more examples.

@urso and @ruflin One of you should review this. I'd like to also add a link to the regexp that we support, but I'm not sure if it's defined as part of Go or what. Please let me know, and I will add a link.

### CODE REVIEW

File: filebeat/docs/configuration.asciidoc [hunk 1]
```diff
@@ -150,48 +151,59 @@ The buffer size every harvester uses when fetching the file. The default is 1638
 
 ===== max_bytes
 
-Maximum number of bytes a single log event can have. All bytes after max_bytes are discarded and not sent.
-This is especially useful for multiline log messages which can get large. The default is 10MB (10485760).
+The maximum number of bytes that a single log event can have. All bytes after `max_bytes` are discarded and not sent.
```
Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: More accurate would probably be "a single log message", as the full event which is sent can be bigger. This is only about the log message (was already wrong before)

**Comment by ruflin** [comment]
LGTM

File: filebeat/docs/configuration.asciidoc [hunk 2]
```diff
@@ -63,27 +63,27 @@ The value that you specify here is used as the `input_type` for each event publi
...
 
 [source,yaml]
 -------------------------------------------------------------------------------------
 include_lines: ["^ERR", "^WARN"]
 -------------------------------------------------------------------------------------
-To include only the lines starting with "ERR" or "WARN".
 
-Note::
-If both `include_lines` and `exclude_lines` are defined, then include_lines is called first. To export all the apache logs except the DBGs, then you can use:
+NOTE: If both `include_lines` and `exclude_lines` are defined, `include_lines` is called first. So, for example, to export all Apache log lines except the debugging messages (DBGs), you can use:
```
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: What's still unclear to me. What's the consequence of having include_lines called first? Does it mean exclude_lines is executed next and my line is still filtered out? Or does it mean the line is forwarded, cause include_lines has higher precedence?

File: filebeat/docs/configuration.asciidoc (hunk 1 above)
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: single log message = maximu size (in bytes?) of the events "message" field (event["message"])

File: filebeat/docs/configuration.asciidoc [hunk 3]
```diff
@@ -150,48 +151,59 @@ The buffer size every harvester uses when fetching the file. The default is 1638
 
 ===== max_bytes
 
-Maximum number of bytes a single log event can have. All bytes after max_bytes are discarded and not sent.
-This is especially useful for multiline log messages which can get large. The default is 10MB (10485760).
+The maximum number of bytes that a single log event can have. All bytes after `max_bytes` are discarded and not sent.
+This setting is especially useful for multiline log messages, which can get large. The default is 10MB (10485760).
```
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: just like a 10MB log line is not large enough ;)

File: filebeat/docs/configuration.asciidoc [hunk 4]
```diff
@@ -150,48 +151,59 @@ The buffer size every harvester uses when fetching the file. The default is 1638
...
 
 ===== multiline
 
-Mutiline can be used for log messages spanning multiple lines. This is common for Java Stack Traces
-or C-Line Continuation. The following example combines all lines following a line start with a `[`.
-This could for example be a Java Stack Trace.
+Options that control how Filebeat deals with log messages that span multiple lines. Multiline messages are common in Java stack traces and C line continuation.
```
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: continuations.

What's C line continuations in log files, though? Maybe just mention "Java stack traces".

File: filebeat/docs/configuration.asciidoc [hunk 5]
```diff
@@ -150,48 +151,59 @@ The buffer size every harvester uses when fetching the file. The default is 1638
...
+This setting is especially useful for multiline log messages, which can get large. The default is 10MB (10485760).
 
 ===== multiline
 
-Mutiline can be used for log messages spanning multiple lines. This is common for Java Stack Traces
-or C-Line Continuation. The following example combines all lines following a line start with a `[`.
-This could for example be a Java Stack Trace.
+Options that control how Filebeat deals with log messages that span multiple lines. Multiline messages are common in Java stack traces and C line continuation. 
+
+The following example shows how to configure Filebeat to handle a multiline message where the first line of the message begins with a bracket (`[`).
```
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: I'm not sure this is a good example. Filter after/match always compare any two consecutive line. Plus after matches, if second line (or subsequent lines) start with `^\[` if I'm not mistaken right now.

Comment (urso) [inline filebeat/docs/configuration.asciidoc]: Filter is okay, my fault. But maybe a better (more generic and less error prone) filter would be based on whitespace, not requiring negate

Comment (dedemorton) [inline filebeat/docs/configuration.asciidoc]: I'm not following your comment. The following config options should process the message in the example, no? 
multiline:
    pattern: ^\[
    negate: true
    match: after

Do you have a better example to recommend? I think seeing an actual example helps users understand the configuration settings better.

**Comment by urso** [comment]
@dedemorton good link to regex support doesn't really exist. There is mostly the [api doc for CompilePOSIX](https://golang.org/pkg/regexp/#CompilePOSIX) and the [syntax doc](https://github.com/google/re2/wiki/Syntax).

File: filebeat/docs/configuration.asciidoc (hunk 5 above)
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: filter should work, but expecting people to copy'n paste this, they will not be happy. Normally log files are a mix of arbitrary logs mixed with java (or some other languages) stack traces.

For example this (not checked this works or '\s' is supported), should be more reliable:

    multiline:
        pattern: \s+
        match: after

Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: @urso @dedemorton For the example, I tried to come up with a very simple example. I think the goal should be that someone that reads the docs here, understands the concept and can apply it to his specific need. We probably should have a subpage (as discussed before) where we provide specific examples and can also mention the edge cases.

File: filebeat/docs/configuration.asciidoc (hunk 2 above)
Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: @urso is for you the text unclear or the behaviour? It means the first one you described. I quite like the example. Perhaps we should use "executed" instead of "called"?

File: filebeat/docs/configuration.asciidoc (hunk 5 above)
Comment (dedemorton) [inline filebeat/docs/configuration.asciidoc]: @ruflin @urso +1 for adding a subpage, though for this initial pass, I wanted to get the basics right first. 
 
I'd like to provide at least 3 or 4 examples of some common use cases that are a bit more complex than the simple ones that I've seen so far. If someone on the dev team can provide me with some example log lines + the config settings for these examples, I can put them into a topic that talks about configuring support for multiline log messages. If you can give me the code examples, I should be able to craft some wording around them.

Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: One other note: How close should we keep this to our LS docs? https://www.elastic.co/guide/en/logstash/current/plugins-codecs-multiline.html

About the examples: +1 We should probably take the examples from the tests with these we are 100% sure they work :-)

Comment (dedemorton) [inline filebeat/docs/configuration.asciidoc]: @ruflin That's a good question. I looked at the Logstash doc when I was working the edits, but I felt that the structure of their doc is so different from our doc that trying to follow their model wouldn't really work for us right now. Also, the language they use in their descriptions is not very precise. For example, they say: "The pattern should match what you believe to be an indicator that the field is part of a multi-line event." This isn't strictly true because not every line in the multiline message matches the pattern depending on what you've specified for other options. So rather than be consistently confusing, I decided to come up with my own descriptions that (I hope) are more precise. Also, because we've renamed a couple of the options, the Logstash descriptions wouldn't make sense in our doc. 

Examples are definitely key here. Can you point me to the tests where we have some examples that I can use?

**Comment by dedemorton** [comment]
@urso @ruflin Please take a look at my changes and merge if they look good.

I think I've addressed your concerns. I went ahead and added the more complex example at the end of the section so it's at least documented until I've had a chance to create another topic that provides more extensive examples. I realize that this page is getting VERY long, but keep in mind that I'm planning to do some restructuring (soon!).

**Comment by monicasarbu** [comment]
LGTM
